### PR TITLE
feature3 added - delete users + interests model

### PR DESCRIPTION
**Pull request by krzysiekpom**
master + feature1 + feature2 + admin to delete users from list & interests model

admin@mail.pl, pass: secret
user@mail.pl, pass: secret

### CODE REVIEW

**Comment by socjopata** [comment]
Same story. Please read my comment on the PR#2

**Comment by krzysiekpom** [comment]
Base branch changed to feature2

File: app/models/user.rb [hunk 1]
```diff
@@ -1,3 +1,10 @@
 class User < ApplicationRecord
   devise :database_authenticatable, :rememberable,  :validatable
+
+  has_many :interests, dependent: :destroy
+ 
+  validates :age, presence: true, numericality: { only_integer: true }, 
+                  inclusion: { in: 8..99 }
+  validates :gender, presence: true, inclusion: { in: ["M","F"] }
+  validates :admin, inclusion: { in: [true, false] }
```
Comment (socjopata) [inline app/models/user.rb]: It's actually better to have a database constraint as `default: false`, if we don't want to end with nil values

File: app/models/user.rb [hunk 2]
```diff
@@ -1,3 +1,10 @@
 class User < ApplicationRecord
   devise :database_authenticatable, :rememberable,  :validatable
+
+  has_many :interests, dependent: :destroy
+ 
+  validates :age, presence: true, numericality: { only_integer: true }, 
+                  inclusion: { in: 8..99 }
+  validates :gender, presence: true, inclusion: { in: ["M","F"] }
```
Comment (socjopata) [inline app/models/user.rb]: Consider extracting these strings to constants like `MALE = "M".freeze` and use them here

File: spec/controllers/users_controller_spec.rb [hunk 3]
```diff
@@ -5,18 +5,37 @@
...
    @user = create(:user)
+   @admin = create(:admin)
   end
 
-  it "displays login page if user notlogged in" do
-    get :index
-    expect(response.response_code).to eql(302)
-    expect(response).to redirect_to(new_user_session_url)
+  context "renders page depending on user being signed in" do
+    it "displays login page if user notlogged in" do
```
Comment (socjopata) [inline spec/controllers/users_controller_spec.rb]: `not logged`

File: spec/controllers/users_controller_spec.rb [hunk 4]
```diff
@@ -5,18 +5,37 @@
...
+      get :index
+      expect(response.response_code).to eql(302)
+      expect(response).to redirect_to(new_user_session_url)
+    end
+
+    it "displays login page if user not logged in" do
+      sign_in @user
+      expect(:get => "/").to route_to("users#index")
+    end
```
Comment (socjopata) [inline spec/controllers/users_controller_spec.rb]: Can you please add a spec, where admin tries to delete himself? Also as an admin user, what kind of UI behavior would you like to experience in a web browser, if you accidentally tried to delete yourself and the server returned a response?

File: spec/factories.rb [hunk 5]
```diff
@@ -1,4 +1,9 @@
 FactoryGirl.define do
+  factory :interest do
+    name "MyString"
+    type ""
```
Comment (socjopata) [inline spec/factories.rb]: Is this factory working at all? Do you have a `type` defined? I saw there is a field named `interest_type` but it's not equal to `type`

File: spec/models/interest_spec.rb [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    end
+  end
+
+  context "name" do
+    it "is present" do
+      @interest.name = " "
+      expect(@interest.valid?).to be false
+    end
+
+    it "is no longer than 255 chars" do
```
Comment (socjopata) [inline spec/models/interest_spec.rb]: Yeah, I hinted about it in a previous PR. Usually before (And after) you write a spec, you should think about the value it brings. The spec should give you confidence that something is working, and that confidence should be worth more than a cost of writing such spec in a first place, and maintaining it later. Here you're basically not testing your code, but you're testing the framework you're using or the database engine itself.

You can assume that Rails is properly tested (if it wasn't, why would you use it in a first place?). Consider deleting this spec.

File: spec/models/user_spec.rb [hunk 7]
```diff
@@ -1,16 +1,51 @@
...
+      @user.age = "test"
+      expect(@user.valid?).to be false
+    end
+
+    it "age is present" do
+      @user.age = "  "
+      expect(@user.valid?).to be false
+    end
+
+    it "gender only F or M value" do
```
Comment (socjopata) [inline spec/models/user_spec.rb]: Similar case. You are testing the Rails framework, not your code here :)

File: spec/models/interest_spec.rb (hunk 6 above)
Comment (krzysiekpom) [reply spec/models/interest_spec.rb]: Agree, this is definitely something I have problems wrapping my head around. I read before about testing the functionalities and not framework, but have a hard time finding out when I'm doing it wrong. In my head it kind of contradicts itself - I try to make a BDD approach, so the test in this case is written before there is a validation on the model - the test will pass only once I add the validations. But as you said, once it's done it seems unnecessary, maybe validations should be added before tests? Will be happy to talk about it.

Comment (socjopata) [reply spec/models/interest_spec.rb]: If writing these specs is helpful for you during the TDD, do so. But don't be afraid to consider some specs as "throw-away-specs" and delete them afterwards, if you don't think the brought a lot of value for finished code. So make it easy for yourself to write code. And remember to make it easy for your "future self" to maintain that code later :)

File: app/models/user.rb (hunk 2 above)
Comment (krzysiekpom) [reply app/models/user.rb]: Done. I put them directly in the model file as constants.rb file seemed like an overkill (it loads constants for the whole application afaik).

File: app/models/user.rb (hunk 1 above)
Comment (krzysiekpom) [reply app/models/user.rb]: `default:false` was already defined in the add_admin migration file. Does it mean I should delete this validation?

File: spec/controllers/users_controller_spec.rb (hunk 4 above)
Comment (krzysiekpom) [reply spec/controllers/users_controller_spec.rb]: I handled this by using rescue_from on exception raised by cancan (displays flash message). Part of the requirements is for the trash can to not display next to admin account, therefore the functionality can't be easily tested through the browser. Easiest way I think is to remove the conditional from _user partial. I added a spec as well.

File: spec/factories.rb (hunk 5 above)
Comment (krzysiekpom) [reply spec/factories.rb]: It isn't - I originally planned to use this, but never did so the error never came up. I removed the whole thing (:interest factory).

File: spec/models/interest_spec.rb (hunk 6 above)
Comment (krzysiekpom) [reply spec/models/interest_spec.rb]: All interest specs were testing the framework, so I deleted the whole file.

File: spec/models/user_spec.rb (hunk 7 above)
Comment (krzysiekpom) [reply spec/models/user_spec.rb]: Same as interest specs - deleted the file.

**Comment by krzysiekpom** [comment]
I implemented the changes. If I had any doubts, I added them in the comments. Thanks for the remarks, these are all very helpful.

File: app/models/user.rb (hunk 2 above)
Comment (socjopata) [reply app/models/user.rb]: I meant to define them as constants, just after the class definition, as good practices suggest to do so. Have a read: http://rubylearning.com/satishtalim/ruby_constants.html

File: spec/controllers/users_controller_spec.rb (hunk 4 above)
Comment (socjopata) [reply spec/controllers/users_controller_spec.rb]: Yeah, you don't test in a browser. You test by writting specs. So we are sure that everything behaves as expected, even a malicious user sends well prepared curl request to our backend. 

And I like displaying a meaningful flash message to the user.

Comment (socjopata) [reply spec/controllers/users_controller_spec.rb]: Can you please point me to that additional spec you wrote?

File: app/models/user.rb (hunk 1 above)
Comment (socjopata) [reply app/models/user.rb]: Yeah, please do :)

File: spec/controllers/users_controller_spec.rb (hunk 4 above)
Comment (krzysiekpom) [reply spec/controllers/users_controller_spec.rb]: Ah sure, I thought you might want to see this in action as you mentioned about UI.

**Comment by socjopata** [comment]
Next time please consider squishing commits, using the `rebase` command. It brings little value for future developers to prevent the history of merging and reversions while introducing the code to the master

**Pull request merged**